kore/src/Kore/Step/Step.hs

```diff
@@ -155,6 +156,9 @@ class UnifyingRule rule where
         -> rule variable1
         -> rule variable2
 
+    -- | Axiom attributes for this rule
+    ruleAttributes :: rule variable -> Attribute.Axiom TermLike.Symbol variable
```

Comment:
This method ties us to rules that have the same representation. Instead, let's add a method like:

```.hs
checkInstantiation :: rule variable -> Map (UnifiedVariable variable) (TermLike variable) -> [InstantiationFailure variable]
```

where `InstantiationFailure` is a sum type of the ways that instantiating the rule can fail. By not tying us to a particular representation of rules, this gives us more freedom in the future. We are also separating concerns by extracting this logic from `EquationalStep`.